### PR TITLE
Create new Mlflow run by default and introduce `run_group`

### PR DESCRIPTION
**Pull request by chenmoneygithub**
# What does this PR do?

Create new Mlflow run by default and introduce `run_group`. Currently the default behavior is to resume the Mlflow run by searching for existing run with the same name. As we discussed we should make the default behavior start a new run. Additionally to mirror the `group` in wandb, we are introducing the `run_group` tag.

# What issue(s) does this change relate to?

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/loggers/mlflow_logger.py [hunk 1]
```diff
@@ -96,13 +98,15 @@ def __init__(
         self.synchronous = synchronous
         self.log_system_metrics = log_system_metrics
         self.ignore_metrics = [] if ignore_metrics is None else ignore_metrics
-        self.ignore_hyperparameters = [] if ignore_hyperparameters is None else ignore_hyperparameters
+        self.ignore_hyperparameters = ([] if ignore_hyperparameters is None else ignore_hyperparameters)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:101]: why this change?

File: composer/loggers/mlflow_logger.py [hunk 2]
```diff
@@ -96,13 +98,15 @@ def __init__(
...
         if self.model_registry_uri == 'databricks-uc':
             if len(self.model_registry_prefix.split('.')) != 2:
                 raise ValueError(
                     f'When registering to Unity Catalog, model_registry_prefix must be in the format ' +
                     f'{{catalog_name}}.{{schema_name}}, but got {self.model_registry_prefix}',
                 )
+        self.run_group = run_group
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:108]: same minor nit: nice to have next to `self.run_name`

File: composer/loggers/mlflow_logger.py [hunk 3]
```diff
@@ -133,14 +137,56 @@ def __init__(
             if env_exp_id is not None:
                 self._experiment_id = env_exp_id
             else:
-                exp_from_name = self._mlflow_client.get_experiment_by_name(name=self.experiment_name)
+                exp_from_name = self._mlflow_client.get_experiment_by_name(name=self.experiment_name,)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:140]:
```suggestion
                exp_from_name = self._mlflow_client.get_experiment_by_name(name=self.experiment_name)
```
Unnecessary change?

File: composer/loggers/mlflow_logger.py [hunk 4]
```diff
@@ -193,7 +212,7 @@ def init(self, state: State, logger: Logger) -> None:
             self._experiment_id, self._run_id = mlflow_ids_list
 
     def after_load(self, state: State, logger: Logger) -> None:
-        logger.log_hyperparameters({'mlflow_experiment_id': self._experiment_id, 'mlflow_run_id': self._run_id})
+        logger.log_hyperparameters({'mlflow_experiment_id': self._experiment_id, 'mlflow_run_id': self._run_id},)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:215]:
```suggestion
        logger.log_hyperparameters({'mlflow_experiment_id': self._experiment_id, 'mlflow_run_id': self._run_id})
```

File: composer/loggers/mlflow_logger.py [hunk 5]
```diff
@@ -271,9 +291,10 @@ def register_model(
             ModelVersion: The registered model.
         """
         if self._enabled:
-            full_name = f'{self.model_registry_prefix}.{name}' if len(self.model_registry_prefix) > 0 else name
+            full_name = (f'{self.model_registry_prefix}.{name}' if len(self.model_registry_prefix) > 0 else name)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:294]: unnecessary change?

File: composer/loggers/mlflow_logger.py [hunk 6]
```diff
@@ -295,13 +316,16 @@ def save_model(self, flavor: Literal['transformers', 'peft'], **kwargs):
...
             if flavor == 'transformers':
                 mlflow.transformers.save_model(**kwargs)
             elif flavor == 'peft':
                 import transformers
 
                 # TODO: Remove after mlflow fixes the bug that makes this necessary
-                mlflow.store._unity_catalog.registry.rest_store.get_feature_dependencies = lambda *args, **kwargs: ''  # type: ignore
+                mlflow.store._unity_catalog.registry.rest_store.get_feature_dependencies = (
+                    lambda *args, **kwargs: ''
+                )  # type: ignore
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:328]: unnecessary change?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM! It looks like your local linter may be applying a whole bunch of changes -- can we remove these? diff should be much smaller I think

File: composer/loggers/mlflow_logger.py (hunk 2 above)
Comment (chenmoneygithub) [reply composer/loggers/mlflow_logger.py]: sure, done!

File: composer/loggers/mlflow_logger.py [hunk 7]
```diff
@@ -76,6 +76,8 @@ def __init__(
         log_system_metrics: bool = True,
         ignore_metrics: Optional[List[str]] = None,
         ignore_hyperparameters: Optional[List[str]] = None,
+        run_group: Optional[str] = None,
+        resume: bool = False,
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply composer/loggers/mlflow_logger.py:84]: sure, done!

File: composer/loggers/mlflow_logger.py [hunk 8]
```diff
@@ -60,6 +60,10 @@ class MLFlowLogger(LoggerDestination):
             log system metrics (CPU/GPU/memory/network usage) during training. (default: ``True``)
         ignore_metrics (List[str], optional): A list of glob patterns for metrics to ignore when logging. (default: ``None``)
         ignore_hyperparameters (List[str], optional): A list of glob patterns for hyperparameters to ignore when logging. (default: ``None``)
+        run_group (str, optional): A string to group runs together. (default: ``None``)
+        resume (bool, optional): If ``True``, Composer will search for an existing run tagged with
+            the `run_name` and resume it. If ``False``, Composer will create a new run. (default:
+            ``False``)
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:66]: Can we add something in docstring that says if True but no run is found, a new one will be created?

File: composer/loggers/mlflow_logger.py [hunk 9]
```diff
@@ -139,8 +147,50 @@ def __init__(
...
+            )
+
+            if len(existing_runs) > 0:
+                self._run_id = existing_runs[0].info.run_id
+            else:
+                new_run = self._mlflow_client.create_run(
+                    experiment_id=self._experiment_id,
+                    run_name=self.run_name,
+                )
+                self._run_id = new_run.info.run_id
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:175]:
```suggestion
            else:
                log.debug('Creating a new mlflow run as resume was set to True but no previous run was found.')
                new_run = self._mlflow_client.create_run(
                    experiment_id=self._experiment_id,
                    run_name=self.run_name,
                )
                self._run_id = new_run.info.run_id
```

Comment (mvpatel2000) [reply composer/loggers/mlflow_logger.py:175]: This shouldnt be a warning because a user likely leaves this on (eg how we treat autoresume for checkpoints), but a debug warning will be helpful if it is unexpected

File: composer/loggers/mlflow_logger.py [hunk 10]
```diff
@@ -139,8 +147,50 @@ def __init__(
...
+            # Search for an existing run tagged with this Composer run if `self.resume=True`.
+            assert self._experiment_id is not None
+            existing_runs = mlflow.search_runs(
+                experiment_ids=[self._experiment_id],
+                filter_string=f'tags.run_name = "{state.run_name}"',
+                output_format='list',
+            )
+
+            if len(existing_runs) > 0:
+                self._run_id = existing_runs[0].info.run_id
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:169]:
```suggestion
            if len(existing_runs) > 0:
                self._run_id = existing_runs[0].info.run_id
                log.debug(f'Resuming mlflow run with run id: {self._run_id}`)
```

Comment (mvpatel2000) [reply composer/loggers/mlflow_logger.py:169]: Same as below

**Review by mvpatel2000 — APPROVED** [review]
LGTM! A minor nit for more helpful logs

File: composer/loggers/mlflow_logger.py (hunk 8 above)
Comment (chenmoneygithub) [reply composer/loggers/mlflow_logger.py:66]: sure, done!

File: composer/loggers/mlflow_logger.py [hunk 11]
```diff
@@ -60,6 +60,10 @@ class MLFlowLogger(LoggerDestination):
             log system metrics (CPU/GPU/memory/network usage) during training. (default: ``True``)
         ignore_metrics (List[str], optional): A list of glob patterns for metrics to ignore when logging. (default: ``None``)
         ignore_hyperparameters (List[str], optional): A list of glob patterns for hyperparameters to ignore when logging. (default: ``None``)
+        run_group (str, optional): A string to group runs together. (default: ``None``)
+        resume (bool, optional): If ``True``, Composer will search for an existing run tagged with
+            the `run_name` and resume it, if no existing run is found, a new run will be created.
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:65]:
```suggestion
            the `run_name` and resume it. If no existing run is found, a new run will be created.
```

File: composer/loggers/mlflow_logger.py [hunk 12]
```diff
@@ -139,8 +147,55 @@ def __init__(
...
+                output_format='list',
+            )
+
+            if len(existing_runs) > 0:
+                self._run_id = existing_runs[0].info.run_id
+                logging.debug(f'Resuming mlflow run with run id: {self._run_id}')
+            else:
+                logging.debug(
+                    'Creating a new mlflow run as `resume` was set to True but no previous run was '
+                    'found.',
```
Comment (mvpatel2000) [inline composer/loggers/mlflow_logger.py:174]:
```suggestion
                log.debug(f'Resuming mlflow run with run id: {self._run_id}')
            else:
                log.debug(
                    'Creating a new mlflow run as `resume` was set to True but no previous run was '
                    'found.',
```

**Review by mvpatel2000 — COMMENTED** [review]
more nits
